src/utils/fireDestinationsFactory.js

```diff
@@ -1,45 +1,80 @@
 import isNil from "./isNil";
-import isObject from "./isObject";
 import isNonEmptyString from "./isNonEmptyString";
 import fireImage from "./fireImage";
 
 const fireOnPage = fireImage;
 
-export default ({ iframe, logger }) => {
-  const fireInIframe = ({ url }) => {
+const processedDests = {
+  loaded: [],
+  errored: [],
+  aborted: []
+};
+
+let numberOfDestinations = -1;
+
+export default ({ iframe, logger, destinationsProcessedDeferred }) => {
+  const checkForCompletion = () => {
+    const processed =
+      processedDests.loaded.length +
+      processedDests.errored.length +
+      processedDests.aborted.length;
+
+    if (processed === numberOfDestinations) {
+      logger.log(
+        `All destinations processed. Loaded: ${
+          processedDests.loaded.length
+        }; errored: ${processedDests.errored.length}; aborted: ${
+          processedDests.aborted.length
+        }`
+      );
+      destinationsProcessedDeferred.resolve(processedDests);
```

Comment:
If I'm reading this right, it seems like `fireDestinations` only supports being called a single time. I believe this is the case because `destinationsProcessedDeferred` will be resolved after the first call to `fireDestinations` once all the destinations have fired. A promise can only be resolved once, so if `fireDestinations` were called again, resolving `destinationsProcessedDeferred` again would do nothing.

This spurs a few questions:

Is `fireDestinations` only intended to be called once? If YES, then why do we expose an `end` method from `createDestinations`, when it could handle "ending" internally after `fire` is called and is completed. If NO, then have a single promise that is resolved after the first `fire` call finishes doesn't seem rational. Instead, it would seem like we would want to return a promise from each `fire` call.

Reply:
@Aaronius thanks for the great catch. I took out end, and return a promise from each fire call. I tested manually on sandbox with logger.log and console.log.

Reply:
Okay, so I think we're saying that `fire` is only intended to be called once. If that's the case, I think there's a bunch more we can simplify here. For example, why do we need to call `createDestinations`, then call the `fire` method, then watch the `destinationsProcessedPromise` to be notified when it's done. Why not just create a module that is a function (say, `fireDestinations`) and when the function is called, it returns a promise that's resolved when destination firing is complete, and then goes to work on firing destinations? I'll try to code up a branch to demonstrate what I'm talking about.

Reply:
Here's what I'm trying to say: https://github.com/adobe/alloy/pull/15

Reply:
fire is intended to be called once per onResponse, assuming there is data. I tested this successfully.